---
title: "NanoBlotPackage"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{NanoBlotPackage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Package Installation
If devtools is not installed yet, run ```install.packages("devtools")```  
Then, install NanoBlot from source using ```devtools::install(build_vignettes = TRUE)```  
To load vignette, run ```browseVignettes("NanoBlotPackage")```
To load, package, run
```{r}
library(NanoBlotPackage)
```


## Setting up Environment 
NanoBlotPackage requires some core dependencies in order for the package to run.  
We have created a yml environment file that can be run to load all dependencies located at ```../scripts/nanoblotenv.yml```

The user can create the conda environment either from terminal or directly within R by invoking the system2() command. A terminal command would look like ```conda env create -f ../scripts/nanoblotenv.yml```      

The last step is to add the nanoblot environment into R's system command so that it can call the functions. 
```
old_path <- Sys.getenv("PATH")
Sys.setenv(PATH = paste(old_path, "/anaconda3/envs/nanoblotenv/bin", sep = ":"))
```
Note, the absolute path will change for each user, so make sure to check the path location to your conda environment. You can find this by calling ```conda env list``` in terminal and copy and pasting the conda env list

Since the R system path is reset with every session, should the user wish not to reset the path each time, it is recommended to download the dependencies directly to the user's console through the pacakge installation program. This pertains exclusively to system2() commands, whereas the rest of the packages can be installed within Rstudio through Bioconductor. 

```{r, include=TRUE}
# If your environment was set up correctly, there should not be an error from calling this command. If you have an error with the build, that means that the source was not properly set up. Call Sys.getenv("PATH") to check if nanoblot env is in system path. 
system2("samtools")
```

## Check Integrity 
Checking integrity is a step that should be performed prior to any plotting, to make sure that none of the samples could have potentially been degraded. 

There are some gtf files if you pass into GeneTargets that contain mitochondrial genes, which might produce an error of ```seqlevels(param) not in BAM header```, so in that case you can use the no mitochondrial gtf file as we used in this example. 
```{r}
bamFileList <- Rsamtools::BamFileList(c("./data/example/WT_sorted_merged.bam",
                                        "./data/example/RRP6_sorted_merged.bam"))

YeastCDSGFF <- rtracklayer::import.gff("./data/annotations/saccharomyces_cerevisiae.20210411.CDS.no_mito.gff")

totalCounts <- checkIntegrity(GeneTargets = YeastCDSGFF, BamFiles = bamFileList)
```


## Basic Workflow 
We start by subsetting our example data files to the probe RPL18A   
Note that we store this in a specified temp folder. If a temp folder is not specified, one will be created automatically for the user
```{r, warning=FALSE}
## Creating BamFileList object to store data
OriginalBamFileList <- Rsamtools::BamFileList(c("./data/example/WT_sorted_merged.bam",
																 "./data/example/RRP6_sorted_merged.bam"))
## Then subsetting original bam files based on RPL18A probe
subsetNanoblot(OriginalBamFileList, "./user_input_files/probes.bed", c("RPL18A_Exon1"), tempFilePath = "./temp")
```

We then prepare for plotting by creating the necessary objects such as the BamFileList that contains all the subsetted bam files, the annotation file, as well as the plotInfo table that will be passed to the ```makeNanoblot``` function
``` {r} 
## Create a new BamFileList object using the newly subsetted bam files
subsettedBamFileList <- Rsamtools::BamFileList(c("./temp/WT_sorted_merged_RPL18A_Exon1.bam",
																 "./temp/RRP6_sorted_merged_RPL18A_Exon1.bam"))
## Change the names of the BamFileList for easier visualization
names(subsettedBamFileList) <- c("WT","RRP6")
## Supply annotation file for yeast provided in example annotations
annotation <- "./data/annotations/Saccharomyces_cerevisiae.R64-1-1.107.gtf"
# Creating the plot info data table
WT_RRP6_plot_info <- data.frame(
	SampleID = c("WT","RRP6"),
	SampleLanes = c(1,2),
	SampleColors = c('blue','red')
)
```

Before plotting, we need to normalize based on differential expression using DESeq2. We supply the code below for a DESeq2 normalization, but keep in mind that the datasets are truncated so it will not be an accurate representation of the actual normalization with the full complete sequencing data. 
```{r}
# First converting the subsettedBamFileList object to Nanoblot data
NanoblotDataWTRRP6 <- bamFileListToNanoblotData(subsettedBamFileList)
# Then performing DESeq2 normalization
unnormalizedLocations <- BiocGenerics::path(OriginalBamFileList)
ds_size_factors <- normalizeNanoblotData(NanoblotDataWTRRP6, "differential", unnormalizedLocations, annotation)
```

This is the default plot output. 
```{r}
makeNanoblot(nanoblotData = NanoblotDataWTRRP6, plotInfo = WT_RRP6_plot_info, size_factors = ds_size_factors)
```


This is the ridge plot output. 
```{r}
makeNanoblot(nanoblotData = NanoblotDataWTRRP6, plotInfo = WT_RRP6_plot_info, blotType = "ridge")
```

This is the violin plot output. 
```{r}
makeNanoblot(nanoblotData = NanoblotDataWTRRP6, plotInfo = WT_RRP6_plot_info, blotType = "violin")
```

## RT-PCR Workflow
IMPORTANT NOTE: As of 04/03/2023, the anaconda package repository is not yet updated with samtools v.1.17, thus if you were using NanoBlot with the conda environment, the ```samtools ampliconclip``` function is not bug fixed to work for RT-PCR. We will be updating the conda environment once anaconda has updated their repository, but in the meantime if you wish to use the RT-PCR functionality, users must install samtools themselves by following installation instrucitons on their website at https://www.htslib.org/download/.   

This is an example of the RT-PCR that was used in the paper with the BAG1 gene in human sequencing data. Keep in mind that due to the large size of the human sequencing data, we are choosing to subet already subsetted files based on the BAG1_Exon just to show users what the subset command will look like. Thus, this workflow skips normalization steps since there are no original sequencing files to base normalization off of. 
```{r}
# Creating BamFileList object with already subsetted BAG1 human bam files 
originalBamFileList <- 
	Rsamtools::BamFileList(c(
"./data/example/scr1_BAG1_Exon.bam",
"./data/example/scr2_BAG1_Exon.bam",
"./data/example/upf1_BAG1_Exon.bam",
"./data/example/smg6_BAG1_Exon.bam",
"./data/example/smg7_BAG1_Exon.bam",
"./data/example/smg67_BAG1_Exon.bam"))
# Renaming file names for better subset file name creation
names(originalBamFileList) <- c("scr1","scr2","upf1","smg6","smg7","smg67")
# Subset command that uses the RT-mode since a viewingWindow argument is provided
subsetNanoblot(originalBamFileList, "./user_input_files/probes.bed", c("BAG1_Exon"), viewingWindow = "BAG1_vw", cDNA = TRUE)

# Creating BamFileList object with subsetted files 
subsettedBamFileList <- 
	Rsamtools::BamFileList(c(
"./temp/scr1_BAG1_Exon.bam", "./temp/scr2_BAG1_Exon.bam",
"./temp/upf1_BAG1_Exon.bam", "./temp/smg6_BAG1_Exon.bam",
"./temp/smg7_BAG1_Exon.bam", "./temp/smg67_BAG1_Exon.bam"
	))
# Renaming files for easier sample name matching with NanoblotDataInfo data frame
names(subsettedBamFileList) <- c("scr1","scr2","upf1","smg6","smg7","smg67")

# Converting BamFileList to Nanoblot data
NanoblotData <- bamFileListToNanoblotData(subsettedBamFileList)
# Creating info data frame to be usd in makeNanoblot() function
NanoblotDataInfo <- data.frame(
	SampleID = c("scr1","scr2","upf1","smg6","smg7","smg67"),
	SampleLanes = seq_along(c("scr1","scr2","upf1","smg6","smg7","smg67")),
	SampleColors = rep("black",6))
# Creating Nanoblot visualization
makeNanoblot(nanoblotData = NanoblotData, plotInfo = NanoblotDataInfo)
```


