```{r}
# These are necessary to start the conda environment --> the only thing we need the user to identify is the path to their conda library. Users can get this using which conda in their terminal. 
system2("/Users/kevinxu/opt/anaconda3/bin/conda", 
				args = c("env", "create", "-f", "./scripts/nanoblotenv.yml"))
reticulate::use_condaenv("nanoblotenv")
```

```{r}
# This should run on all computers
TestBamFileList <- Rsamtools::BamFileList(c("./data/example/WT_sorted_merged.bam",
																 "./data/example/RRP6_sorted_merged.bam"))
subsetNanoblot(TestBamFileList, "./user_input_files/probes.bed", c("RPL18A_Exon1"))
```



```{r}
# Shortened workflow, Meant to be run on all computers
TestBamFileList <- Rsamtools::BamFileList(c("./temp/WT_sorted_merged_RPL18A_Exon1.bam",
																 "./temp/RRP6_sorted_merged_RPL18A_Exon1.bam"))
names(TestBamFileList) <- c("WT","RRP6")
annotation <- "./data/annotations/Saccharomyces_cerevisiae.R64-1-1.107.gtf"
unnormalizedLocations <- c("./data/example/WT_sorted_merged.bam", "./data/example/RRP6_sorted_merged.bam")

WT_RRP6Test2 <- data.frame(
	SampleID = c("WT","RRP6"),
	SampleLanes = c(1,2),
	SampleColors = c('blue','red')
)

NanoblotDataWTRRP6 <- bamFileListToNanoblotData(TestBamFileList)
ds_size_factors <- normalizeNanoblotData(NanoblotDataWTRRP6, "differential", unnormalizedLocations, annotation)
makeNanoblot(nanoblotData = NanoblotDataWTRRP6, plotInfo = WT_RRP6Test2, blotType = "ridge")
makeNanoblot(nanoblotData = NanoblotDataWTRRP6, plotInfo = WT_RRP6Test2, blotType = "violin")
makeNanoblot(nanoblotData = NanoblotDataWTRRP6, plotInfo = WT_RRP6Test2, size_factors = ds_size_factors)
```

```{r}
## Testing BAG1 workflow, to be run on Nanopore computer
UnnormalizedBamFileList <- 
	Rsamtools::BamFileList(c(
"/media/guillaume-chanfreau/HD3/human_NMD_data/scr1/sorted_merged.bam",
"/media/guillaume-chanfreau/HD3/human_NMD_data/scr2/sorted_merged.bam",
"/media/guillaume-chanfreau/HD3/human_NMD_data/UPF1/sorted_merged.bam",
"/media/guillaume-chanfreau/HD3/human_NMD_data/SMG6/sorted_merged.bam",
"/media/guillaume-chanfreau/HD3/human_NMD_data/SMG7/sorted_merged.bam",
"/media/guillaume-chanfreau/HD3/human_NMD_data/SMG6SMG7/sorted_merged.bam"))
names(UnnormalizedBamFileList) <- c("scr1","scr2","upf1","smg6","smg7","smg67")
subsetNanoblot(UnnormalizedBamFileList, "/home/guillaume-chanfreau/Desktop/user_input_files/probes.bed", c("BAG1_Exon"), viewingWindow = "BAG1_vw", cDNA = TRUE, RTPCR = TRUE)

annotation <- "/home/guillaume-chanfreau/Downloads/Homo_sapiens.GRCh38.107.gtf"
unnormalizedLocations <-as.vector(BiocGenerics::path(UnnormalizedBamFileList))
SubsetBamFileList <- 
	Rsamtools::BamFileList(c(
"./temp/scr1_BAG1_Exon.bam", "./temp/scr2_BAG1_Exon.bam",
"./temp/upf1_BAG1_Exon.bam", "./temp/smg6_BAG1_Exon.bam",
"./temp/smg7_BAG1_Exon.bam", "./temp/smg67_BAG1_Exon.bam"
	))
names(SubsetBamFileList) <- c("scr1","scr2","upf1","smg6","smg7","smg67")

NanoblotData <- bamFileListToNanoblotData(SubsetBamFileList)
NanoblotDataInfo <- data.frame(
	SampleID = c("scr1","scr2","upf1","smg6","smg7","smg67"),
	SampleLanes = seq_along(c("scr1","scr2","upf1","smg6","smg7","smg67")),
	SampleColors = rep("black",6))
ds_size_factors <- normalizeNanoblotData(NanoblotData, "differential", unnormalizedLocations, annotation)
makeNanoblot(nanoblotData = NanoblotData, plotInfo = NanoblotDataInfo, size_factors = ds_size_factors)

```


```{r}
# To be run only on Nanopore computer --> checking DESeq2 with replicate samples 
nanoblotData <- data.frame(SampleID = c())
annotation <- "./user_input_files/Saccharomyces_cerevisiae.R64-1-1.107.gtf"
unnormalizedLocations <- c("/media/guillaume-chanfreau/HD3/KX_Reps/ERR5767189/WT_rep2_mapped.sorted.bam",
													 "/media/guillaume-chanfreau/HD3/KX_Reps/ERR5767190/KO_rep1_mapped.sorted.bam",
													 "/media/guillaume-chanfreau/HD3/KX_Reps/ERR5767191/KO_rep2_mapped.sorted.bam",
													 "/media/guillaume-chanfreau/HD3/KX_Reps/ERR5767192/WT_rep3_mapped.sorted.bam",
													 "/media/guillaume-chanfreau/HD3/KX_Reps/ERR5767193/KO_rep3_mapped.sorted.bam")

coldata <- data.frame(condition = c("WT","KO","KO","WT","KO"))
rownames(coldata) <- c("WT_rep2_mapped.sorted.bam","KO_rep1_mapped.sorted.bam",
											 "KO_rep2_mapped.sorted.bam","WT_rep3_mapped.sorted.bam",
											 "KO_rep3_mapped.sorted.bam")

ds_size_factors <- normalizeNanoblotData(nanoblotData, "differential", unnormalizedLocations, annotation, coldata,  "/home/guillaume-chanfreau/Desktop/WT_KO_DESeq2.csv")


```

```{r}
# To be only run on Nanopore computer --> calculating integrity
TestBamFileList <- Rsamtools::BamFileList(c("/media/guillaume-chanfreau/HD3/TRAMP_Direct/WT.bam", "/media/guillaume-chanfreau/HD3/TRAMP_Direct/RRP6.bam", "/media/guillaume-chanfreau/HD3/TRAMP_Direct/TRF4_3.bam"))

YeastCDSGFF <- rtracklayer::import.gff("/home/guillaume-chanfreau/Annotation_files/saccharomyces_cerevisiae.20210411.CDS.no_mito.gff")

checkIntegrity(GeneTargets = YeastCDSGFF, BamFiles = TestBamFileList)
```

```{r}
# To be only run on Kevin's computer 
YeastCDSGFF <- rtracklayer::import.gff("/home/guillaume-chanfreau/Annotation_files/saccharomyces_cerevisiae.20210411.CDS.no_mito.gff")
TestBamFileList <- Rsamtools::BamFileList(c("./data/example/WT_sorted_merged.bam",
                                            "./data/example/RRP6_sorted_merged.bam"))

counts <- checkIntegrity(GeneTargets = YeastCDSGFF, BamFiles = TestBamFileList)
```



